lib/ansible/modules/packaging/language/pip.py

```diff
@@ -234,6 +252,53 @@
 _SPECIAL_PACKAGE_CHECKERS = {'setuptools': 'import setuptools; print(setuptools.__version__)',
                              'pip': 'import pkg_resources; print(pkg_resources.get_distribution("pip").version)'}
 
+_VCS_RE = re.compile(r'(svn|git|hg|bzr)\+')
+
+op_dict = {">=": operator.ge, "<=": operator.le, ">": operator.gt,
+           "<": operator.lt, "==": operator.eq, "!=": operator.ne, "~=": operator.ge}
+
+
+def _is_vcs_url(name):
+    """Test whether a name is a vcs url or not."""
+    return re.match(_VCS_RE, name)
+
+
+def _is_valid_package_name(name):
+    """Test whether a name is package name or version specifier."""
+    return not name.lstrip().startswith(tuple(op_dict.keys()))
+
+
+def _recover_package_name(names):
+    """Recover package names as list from user's raw input.
+
+    :input: a mixed and invalid list of names or version specifiers
+    :return: a list of valid package name
+
+    eg.
+    input: ['django>1.11.1', '<1.11.3', 'ipaddress', 'simpleproject>1.1.0', '<2.0.0']
+    return: ['django>1.11.1,<1.11.3', 'ipaddress', 'simpleproject>1.1.0,<2.0.0']
+
+    input: ['django>1.11.1,<1.11.3,ipaddress', 'simpleproject>1.1.0,<2.0.0']
+    return: ['django>1.11.1,<1.11.3', 'ipaddress', 'simpleproject>1.1.0,<2.0.0']
+    """
+    # rebuild input name to a flat list so we can tolerate any combination of input
+    tmp = []
+    for one_line in names:
+        tmp = tmp + one_line.split(",")
+    names = tmp
```

Comment:
This whole block could've been done with a nested list comprehension without any need for temporary variable introduction:
```python
names = [
    name
    for name_list in names
        for name in name_list.split(',')
]
```

Alternative FP approach would look like:
```python
from itertools import partial, reduce
...
names = reduce(
    lambda l, e: l + e,
    map(
        partial(str.split, sep=','),
        names
    ),
    [],
)
```
(but Pythonistas don't like such style in general)

Reply:
I used this at the beginning, but Toshio thinks it's better to use a for loop so it's easy to read.